src/LiveComponent/src/EventListener/LiveComponentSubscriber.php

```diff
@@ -160,22 +159,29 @@ public function onKernelController(ControllerEvent $event): void
             throw new NotFoundHttpException(sprintf('The action "%s" either doesn\'t exist or is not allowed in "%s". Make sure it exist and has the LiveAction attribute above it.', $action, \get_class($component)));
         }
 
-        if ($request->attributes->has('_mounted_component')) {
-            // sub-request
-            $event->setController([$request->attributes->get('_mounted_component')->getComponent(), $action]);
-        } else {
-            $data = $this->parseDataFor($request);
-
-            $request->attributes->set('_component_action_args', $data['args']);
+        /*
+         * Either we:
+         *      A) To not have a _mounted_component, so hydrate $component
+         *      B) We DO have a _mounted_component, so no need to hydrate,
+         *          but we DO need to make sure it's set as the controller.
+         */
+        if (!$request->attributes->has('_mounted_component')) {
```

Comment:
Once I dug in, I better understood the limitations and complexities that you were dealing with. This is really the only significant change. I wanted the "role" of `onKernelController` to be more straightforward - vs being the "traffic cop" that has whatever code is necessary to get the job done.

The job of `onKernelController` is now to do the following, and it only "acts" when the controller will be a method on the component (i.e. it's skipped for `_batch`):

A) Either mount the component (which is already set as the controller object. Or, if already mounted, make sure the mounted component IS the controller object.

B) Figure out the arguments to pass to the controller.

That feels more straightforward to me.